docs/architecture/adr-054-semver-compatible-modules.md

```diff
@@ -0,0 +1,408 @@
+mitigation.
+
+### Problem 3: Handling Minor Version Incompatibilities
+
+Imagine that we solve the first two problems but now have a scenario where
+`bar/v2` wants the option to use `MsgDoSomething.condition` which only `foo/v2`
+supports. If `bar/v2` works with `foo` `v1` and sets `condition` to some non-nil
+value, then `foo` will silently ignore this field resulting in a silent logic
+possibly dangerous logic error. If `bar/v2` were able to check whether `foo` was
+on `v1` or `v2` and dynamically, it could choose to only use `condition` when
+`foo/v2` is available. Even if `bar/v2` were able to perform this check, however,
+how do we know that it is always performing the check properly. Without
+some sort of framework-level [unknown field filtering](./adr-020-protobuf-transaction-encoding.md#unknown-field-filtering),
+it is hard to know whether these pernicious hard to detect bugs are getting into
+our app and a client-server layer such as [ADR 033: Inter-Module Communication](./adr-033-protobuf-inter-module-comm.md)
+may be needed to do this.
+
+### Solutions
+
+### Approach A) Separate API and State Machine Modules
+
+One solution (first proposed in https://github.com/cosmos/cosmos-sdk/discussions/10582) is to isolate all protobuf generated code into a separate module
+from the state machine module. This would mean that we could have state machine
+go modules `foo` and `foo/v2` which could use a types or API go module say
+`foo/api`. This `foo/api` go module would be perpetually on `v1.x` and only
+accept non-breaking changes. This would then allow other modules to be
+compatible with either `foo` or `foo/v2` as long as the inter-module API only
+depends on the types in `foo/api`. It would also allow modules `foo` and `bar`
+to depend on each other in that both of them could depend on `foo/api` and
+`bar/api` without `foo` directly depending on `bar` and vice versa.
+
+This is similar to the naive mitigation described above except that it separates
+the types into separate go modules which in and of itself could be used to
+break circular module dependencies. It has the same problems as the naive solution,
+otherwise, which we could rectify by:
+1. removing all state machine breaking code from the API module (ex. `ValidateBasic` and any other interface methods)
+2. embedding the correct file descriptors for unknown field filtering in the binary
+
+#### Migrate all interface methods on API types to handlers 
+
+To solve 1), we need to remove all interface implementations from generated
+types and instead use a handler approach which essentially means that given
+a type `X`, we have some sort of resolver which allows us to resolve interface
+implementations for that type (ex. `sdk.Msg` or `authz.Authorization`). For
+example:
+
+```go
+func (k Keeper) DoSomething(msg MsgDoSomething) error {
+	var validateBasicHandler ValidateBasicHandler
+	err := k.resolver.Resolve(&validateBasic, msg)
```

Comment:
Where would resolver handlers be implemented? The state machine module, api module, or neither?

Reply:
Yes. They would be registered similar to how we do RegisterInterfaces or RegisterServices

Reply:
How is this different from what we currently have per-separation of API/state machine module where the method receiver for each type message is in the "state machine module"?

Reply:
It just wouldn't be defined directly on the protobuf type anymore. Although in the end maybe it wouldn't be too different. What we're doing with Any unpacking is pretty hacks and should go away